text/0000-thread-safe-handles.md

```diff
@@ -0,0 +1,362 @@
+        println!("{:?}", cb);
+    });
+
+    cx.undefined()
+}
+```
+
+While `Persistent<_>` may be shared across threads, the inner contents can only be accessed on the main thread.
+
+```rust
+fn log(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let persistent_cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+    let cb = persistent_cb.into_inner(&cx);
+
+    println!("{:?}", cb);
+
+    cx.undefined()
+}
+```
+
+`Persistent<_>` handles that are not unwrapped must be manually dropped. If a `Persistent<_>` is not consumed, neon will `panic` when the value is dropped.
+
+```rust
+fn log(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let persistent_cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+
+    // `Persistent` are reference counted and clone-able.
+    let cb = persistent_cb.clone().into_inner(&cx);
+
+    // The callback is no longer needed
+    cb.drop(&cx);
+
+    cx.undefined()
+}
+```
+
+### `neon::handle::EventQueue`
+
+Once a value is wrapped in a `Persistent<_>` handle, it must be sent back to the main JavaScript thread to unwrap. `EventQueue` provides a mechanism for requesting work be peformed on the main thread.
+
+To schedule work, send a closure to the event queue:
+
+```rust
+fn thread_callback(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+    let queue = EventQueue::new(&mut cx)?;
+
+    std::thread::spawn(move || {
+        queue.send(move |mut cx| {
+            let this = cx.undefined();
+            let msg = cx.string("Hello, World!");
+            let cb = cb.into_inner(&cx);
+
+            cb.call(&mut cx, this, vec![msg]).unwrap();
+        });
+    });
+
+    cx.undefined()
+}
+```
+
+`Persistent<_>` are clone-able and can be used many times. `EventQueue` are `Sync` and may be shared across threads.
+
+```rust
+fn thread_callback(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+    let queue = Arc::new(EventQueue::new(&mut cx)?);
+
+    let handles = (1..=10).into_iter().map(|i| {
+        let queue = Arc::clone(queue);
+        let cb = cb.clone();
+
+        std::thread::spawn(move || {
+            queue.send(move |mut cx| {
+                let this = cx.undefined();
+                let msg = cx.string(format!("Count: {}", i));
+                let cb = cb.into_inner(&cx);
+
+                cb.call(&mut cx, this, vec![msg]).unwrap();
+            });
+        });
+    });
+
+    // Each thread has a clone; drop the original.
+    cb.drop(&cx);
+
+    cx.undefined()
+}
+```
+
+Instances of `EventQueue` will keep the event loop running and prevent the process from exiting. The `EventQueue::unref` method is provided to change this behavior and allow the process to exit while an instance of `EventQueue` still exists.
+
+However, calls to `EventQueue::schedule` _might_ not execute before the process exits.
```

Comment:
Do you mean they might not execute before the process exits specifically after calling `unref()`? It would be clearer to be explicit here.

Reply:
Is `unref()` the name N-API uses? It's a little awkward. I don't feel strongly about it. I wonder if an English verb like `detach()` or `release()` might read better.

Reply:
Oh I see, it's paired with `ref()`. Probably ignore this suggestion. :)

Reply:
@dherman `unref` means that this `EventQueue` won't keep the event loop alive. If the event loop otherwise becomes empty and ends, a successfully queued event might not execute. I think `unref` is a good name because it is the idiomatic term used in JavaScript/Node.

https://nodejs.org/api/timers.html#timers_immediate_unref